tools/cli/main.cpp

```diff
@@ -12,246 +12,859 @@
-    auto it = args->find(arg);
-    return it != args->end() && it->second.asBool();
+#include "argtable/argtable3.h"
+#include <iomanip>
+
+
+#define iREPORT_EXIT(x)\
+    err = (x); \
+    if (err) { \
+        goto cleanup; \
+    } \
+
+#define iREPORT_EXIT_MSG(x, msg)\
+    err = (x); \
+    if (err) { \
+        std::cerr << msg << std::endl; \
+        goto cleanup; \
+    } \
+
+#define iFAIL(code)\
+    err = (code); \
+    goto fail; \
+
+
+#define EXIT(STATE) \
+    err = (STATE); \
+    goto exit; \
+
+#define iFAIL_MSG(code, msg) \
+    std::cerr << msg << std::endl; \
+    iREPORT_EXIT(code) \
+
+#define PRINT_ION_COMMAND printf("\t%s", ION_CLI_PNAME)
+#define PRINT_HEADER_OPTIONS std::cout << "Options: " << std::endl << std::endl
+#define PRINT_HEADER_GLOSSARY std::cout << std::endl << "Glossary: " << std::endl << std::endl
+#define PRINT_HEADER_EXAMPLES std::cout << "Examples: " << std::endl << std::endl
+#define PRINT_HEADER_COMMANDS std::cout << "Commands: " << std::endl << std::endl
+
+#define PRINT_SUBCOMMAND_HELP_ON_ERROR(tablename) \
+    PRINT_HEADER_OPTIONS; \
+    PRINT_ION_COMMAND; \
+    arg_print_syntaxv(stdout, tablename, "\n"); \
+    PRINT_HEADER_GLOSSARY; \
+    arg_print_glossary_gnu(stdout, tablename); \
+    PRINT_HEADER_EXAMPLES;\
+
+
+//
+// Constants
+//
+
+// help and version
+static const char *const HELP_SHORT_OPT = "h";
+static const char *const HELP_LONG_OPT = "help";
+static const char *const HELP_COMMAND_NAME = HELP_LONG_OPT;
+static const char *const HELP_GLOSSARY = "print the help message and exit";
+static const char *const VERSION_SHORT_OPT = "v";
+static const char *const SYM_TABLE_VERSION_SHORT_OPT = VERSION_SHORT_OPT;
+static const char *const VERSION_LONG_OPT = "version";
+static const char *const VERSION_COMMAND_NAME = VERSION_LONG_OPT;
+static const char *const VERSION_GLOSSARY = "print the command's version number and exit";
+
+// Max values for
+//   - no errors
+//   - no input for multiple input options
+static const int ARG_MAX_ERRORS = 20;
+static const int MAX_NO_INPUTS = 24;
+
+// Common args shared by subcommands
+static const char *const OUTPUT_SHORT_OPT = "o";
+static const char *const OUTPUT_LONG_OPT = "output";
+static const char *const FILE_DATATYPE = "<file>";
+static const char *const OUTPUT_FILE_GLOSSARY = "Output file (default: stdout).";
+
+static const char *const ERR_SHORT_OPT = "e";
+static const char *const ERR_LONG_OPT = "error-report";
+static const char *const ERR_GLOSSARY = "Error report file (default: stderr).";
+
+static const char *const OUT_TYPE_SHORT_OPT = "f";
+static const char *const OUT_TYPE_LONG_OPT = "output-format";
+static const char *const TYPE_DATATYPE = "<type>";
+static const char *const OUT_TYPE_GLOSSARY1 = "Output format, from the set (text | pretty | binary | events | none).'events' is only available with the 'process' command, and outputs a serialized EventStream representing the input Ion stream(s).[default: pretty]";
```

Comment:
Why are some of these suffixed with "GLOSSARY" while others have "GLOSSARY1"?

Reply:
I originally was using the argtable mechanism that does not wrap or format that required the strings to be split thus the suffix number. Since then I found a better that will auto-format and I missed the naming convention.